### ISSUE TITLE
Cache reads generate more range tombstones than necessary

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#2581]
*Installation details*
Scylla version (or git commit hash): 2.0+

`cache_streamed_mutation` will check for overlapping range tombstones, trim them and emit, for every range between two consecutive entries. Mainly because we don't know if something wasn't inserted into cache, which affects a later range, which is now continuous, but that information wasn't there since we last checked. That may result in a single range tombstone being split and emitted multiple times.

We should optimize this and avoid redundant emissions, for instance by tracking the current active tombstone and not emitting if no new information is present.

**Comment by tgrabiec** [comment]
In addition to that, if large partition is mostly range tombstones, reading becomes quadratic because cache reader makes a vector of all relevant range tombstones before filling the buffer. In the worst case of one partition with only range tombstones, this will copy whole cache on each fill_buffer().

**Comment by vladzcloudius** [comment]
@amihayday FYI

**Comment by tgrabiec** [comment]
It doesn't seem hard to fix. Instead of building a vector of all relevant range tombstones for the requested slice, provide a merged stream of range tombstones and let the consumer signal that iteration should stop (e.g. via provided consumer callback).

**Comment by tgrabiec** [comment]
perf_row_cache_update has a test case which reproduces this, commented out:

```c++
            // Takes a huge amount of time due to https://github.com/scylladb/scylla/issues/2581#issuecomment-398030186,
            // disable until fixed.
            // test_partition_with_lots_of_range_tombstones();

```

**Comment by avikivity** [comment]
We won't backport it. But it's available in 4.6 and 5.0 now.
